### PR TITLE
Update Tree-Sitter C grammar

### LINKED ISSUE #1337: Mainstream Tree-Sitter issues for Rizin use cases
Some issues that are in the C grammar of the Tree-Sitter that are relevant:
- https://github.com/tree-sitter/tree-sitter-c/issues/143
- https://github.com/tree-sitter/tree-sitter-c/issues/59
- https://github.com/tree-sitter/tree-sitter-c/issues/44
- https://github.com/tree-sitter/tree-sitter-cpp/issues/63
- https://github.com/tree-sitter/tree-sitter-c/issues/48
- https://github.com/tree-sitter/tree-sitter-c/issues/49
- https://github.com/tree-sitter/tree-sitter-c/issues/47
- https://github.com/tree-sitter/tree-sitter-c/issues/45
- https://github.com/tree-sitter/tree-sitter-c/issues/42
- https://github.com/tree-sitter/tree-sitter-c/issues/132
- https://github.com/tree-sitter/tree-sitter-c/pull/127
- https://github.com/tree-sitter/tree-sitter-c/issues/80
- https://github.com/tree-sitter/tree-sitter-c/issues/154

### LINKED ISSUE #1666: Support types with multiple declarators
```c
struct bla {
    int foo, bar;
    void *b;
};
union foo {
    float a, b;
    int c, d;
};
typedef struct {
    char *a;
    void *b;
} ts, *pts;
```

See `librz/type/parser/type_parser.c`

The problem that such AST nodes have multiple nodes with the same identifier "declarator". And there seems no the API to get the next named field.
```
struct bla { int foo, bar; };

translation_unit [0, 0] - [1, 0]
  struct_specifier [0, 0] - [0, 28]
    name: type_identifier [0, 7] - [0, 10]
    body: field_declaration_list [0, 11] - [0, 28]
      field_declaration [0, 13] - [0, 26]
        type: primitive_type [0, 13] - [0, 16]
        declarator: field_identifier [0, 17] - [0, 20]
        declarator: field_identifier [0, 22] - [0, 25]
```

Asked the help of the Tree-Sitter community in https://github.com/tree-sitter/tree-sitter/discussions/1382

### PR DESCRIPTION
**Pull request by XVilka**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

As Tree-Sitter itself and C grammar fixed [some of the issues](https://github.com/rizinorg/rizin/issues/1337) that affected Rizin, update them before the release

Partially addresses https://github.com/rizinorg/rizin/issues/3498 (parsing stage)
Partially addresses https://github.com/rizinorg/rizin/issues/2308 (parsing stage)
Fixes https://github.com/rizinorg/rizin/issues/1666
Fixes https://github.com/rizinorg/rizin/issues/1337 (only left is preprocessor, should be handled differently: https://github.com/rizinorg/rizin/pull/3661)

**Test plan**

CI is green

### CODE REVIEW

**Comment by XVilka** [comment]
Waiting for https://github.com/tree-sitter/tree-sitter-c/pull/148 and https://github.com/tree-sitter/tree-sitter-c/pull/146 first, then I will update and regenerate again

**Comment by XVilka** [comment]
The only left unfixed for now is https://github.com/rizinorg/rizin/issues/2190 but this one is harder and better to do separately (likely not in this release either as it depends on the mainstream bug: https://github.com/tree-sitter/tree-sitter-c/issues/126)

File: librz/type/parser/types_parser.c [hunk 1]
```diff
@@ -447,92 +458,37 @@ int parse_struct_node(CParserState *state, TSNode node, const char *text, Parser
...
+
+		// Parse types first
+		free(real_type);
+		real_type = ts_node_sub_string(field_type, text);
+		if (!real_type) {
+			parser_error(state, "ERROR: Struct member type should not be NULL!\n");
+			node_malformed_error(state, child, text, "struct field");
+			result = -1;
+			goto srnexit;
+		}
```
Comment (Rot127) [inline librz/type/parser/types_parser.c:470]: All those repeating tests could probably be moved to a separated function.
So `result  = -1` is moved to the failure label. And replace it with
```c
if (!parse_YYY()) {
   goto srn_fail_exit;
}
```

File: librz/type/parser/types_parser.c [hunk 2]
```diff
@@ -542,39 +498,93 @@ int parse_struct_node(CParserState *state, TSNode node, const char *text, Parser
...
+					result = -1;
+					goto srnexit;
+				}
+				// Add a struct member
+				RzVector *members = &struct_pair->btype->struct_data.members;
+				RzTypeStructMember memb = {
+					.name = membname,
+					.type = rz_type_clone(membtpair->type),
+					.offset = 0, // FIXME
+					.size = 0, // FIXME
```
Comment (Rot127) [inline librz/type/parser/types_parser.c:546]: Maybe describe what needs to be fixed here.

File: librz/type/parser/types_parser.c [hunk 3]
```diff
@@ -542,39 +498,93 @@ int parse_struct_node(CParserState *state, TSNode node, const char *text, Parser
...
+				};
+				void *element = rz_vector_push(members, &memb); // returns null if no space available
```
Comment (Rot127) [inline librz/type/parser/types_parser.c:548]: Comment that it returns NULL on failure can be removed.

File: librz/type/parser/types_parser.c [hunk 4]
```diff
@@ -542,39 +498,93 @@ int parse_struct_node(CParserState *state, TSNode node, const char *text, Parser
...
+					parser_error(state, "Error appending bitfield struct member to the base type\n");
+					result = -1;
+					goto srnexit;
+				}
+			} else if (is_declarator(ts_node_type(field_declarator))) {
+				// 2nd case, normal structure
+				// AST looks like
+				// type: (primitive_type) declarator: (field_identifier)
+				// or type: (primitive_type) declarator: (...) declarator (...)
+				// Augment resulting type field with the data from parsed declarator
```
Comment (Rot127) [inline librz/type/parser/types_parser.c:559]: Maybe not in this PR as well. But moving each case into a separated function should increase readability greatly.

File: librz/type/parser/types_parser.c [hunk 5]
```diff
@@ -542,39 +498,93 @@ int parse_struct_node(CParserState *state, TSNode node, const char *text, Parser
...
+					result = -1;
+					goto srnexit;
+				}
+				// Add a struct member
+				RzVector *members = &struct_pair->btype->struct_data.members;
+				RzTypeStructMember memb = {
+					.name = membname,
+					.type = rz_type_clone(membtpair->type),
+					.offset = 0, // FIXME
+					.size = 0, // FIXME
```
Comment (Rot127) [inline librz/type/parser/types_parser.c:573]: What needs to be fixed?

**Review by Rot127 — APPROVED** [review]
Nothing which was related to this specific change. But the length could be reduced in the future. Should reduce the difficulty changing the code here.

**Pull request merged**